src/devices/GoPiGo3/GoPiGo3.cs

```diff
@@ -0,0 +1,900 @@
+                towait += 1 + inBytes;
+            timeout += (int)(0.115 * towait);
+            // but make sure we wait a minimum of 1 ms
+            if (towait > 0)
+            {
+                timeout = Math.Clamp(timeout, 1, 32);
+                Thread.Sleep((int)timeout);
+            }
+
+            timeout = stopwatch.ElapsedMilliseconds + 5;
+            while (true)
+            {
+                try
+                {
+                    return GetGrooveValue(port);
+                }
+                catch (Exception ex) when (ex is IOException)
+                {
+                    throw new IOException($"{nameof(GrooveI2cTransfer)} error: timeout while transfering the I2C data");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Start an I2C transaction
+        /// </summary>
+        /// <param name="port">The Groove Port, one at the time Groove1 or Groove2</param>
+        /// <param name="addr">The I2C address of the slave to be addressed.</param>
+        /// <param name="arrayToSend">An array of bytes to send.</param>
+        /// <param name="inBytes">The number of bytes to read.</param>
+        public void GrooveI2cStart(GroovePort port, byte addr, byte[] arrayToSend, byte inBytes = 0)
+        {
+            SpiMessageType message_type;
+            byte port_index;
+            if (port == GroovePort.Groove1)
+            {
+                message_type = SpiMessageType.StartGroove1I2c;
+                port_index = 0;
+            }
+            else if (port == GroovePort.Groove2)
+            {
+                message_type = SpiMessageType.StartGroove2I2c;
+                port_index = 1;
+            }
+            else
+            {
+                throw new ArgumentException($"{nameof(GrooveI2cStart)} error: Port unsupported. Must be either Groove 1 or Groove 2.");
+            }
+            var address = ((addr & 0x7F) << 1);
+            if (inBytes > GrooveI2cLengthLimit)
+                throw new ArgumentException($"{nameof(GrooveI2cStart)} error: Read length error. Up to {GrooveI2cLengthLimit} bytes can be read in a single transaction.");
+            if (arrayToSend.Length > GrooveI2cLengthLimit)
+                throw new ArgumentException($"{nameof(GrooveI2cStart)} error:Write length error. Up to {GrooveI2cLengthLimit}  bytes can be written in a single transaction.");
+
+            byte[] outArray = { SpiAddress, (byte)message_type, (byte)address, inBytes, (byte)arrayToSend.Length };
+            Array.Resize(ref outArray, outArray.Length + arrayToSend.Length);
+            Array.Copy(arrayToSend, 0, outArray, outArray.Length - arrayToSend.Length, arrayToSend.Length);
+            var reply = SpiTransferArray(outArray);
+
+            GrooveSensor[port_index].I2cDataLength = inBytes;
+            if (reply[3] != 0xA5)
+                throw new IOException($"{nameof(GrooveI2cStart)} error: No SPI response");
+            if (reply[4] != 0)
```

Comment:
0 should be a const in this context (at least local constant if not used anywhere else)

Reply:
yes, I can, it is used few times to check if the I2C transmitted data are correct